### ISSUE TITLE
ValueType: Discussion on using a header bit for flattened arrays

### ISSUE THREAD

**Issue by tajila** [eclipse/openj9#12158]
Valuetypes is going to introduce covariant arrays to the JDK. This means that when an array instance has a static type of `[java/lang/Object;` or `[com/some/InterfaceOrValueCapableAbstractClass;` there is no guarantee as to what the element size of the array may be if it is flattened. 

Currently, the way to check if an array instance is flattened is by checking the classFlags (`arrayInstance->ramClass->classFlags`). Having a faster path to getting this information by using up a bit in the object header has shown to improve performance. See https://github.com/eclipse/openj9/pull/12105 for the bits.

Currently there are no free bits available in the header. These are the proposed solutions:

1) Overload the bit 4 (Object has been hashed) to denote a flattened array as well. 

2) Use the 32nd bit in the `size` field  of the `J9IndexableObject` header to denote a flattened array.

3) Use one of the low bits in `dataAddr` field of the `J9IndexableObject` header to denote a flattened array

4) Do nothing

**Comment by tajila** [comment]
@0xdaryl @vijaysun-omr @dmitripivkine @hzongaro @0xdaryl @Leonardo2718

**Comment by dmitripivkine** [comment]
@amicic FYI

**Comment by dmitripivkine** [comment]
I don't see how "Overload the bit 4 (Object has been hashed) to denote a flattened array as well" can be real solution. 
If GC will maintain FA as always "hashed" (grow by hash code unconditionally if moved) - this might be not even possible because GC operates with "hashed" and "moved" bits as a pair (enum) and all three non-zero combinations are used (10,11,01), so "hashed" bit might be suddenly be set to 0 by GC
If GC will need to perform check for FA for **each hashed object** is going to be a huge performance impact. And how we are going to recognize hashed FA objects in this case? Essentially it means each FA object has to be hashed at the moment of creation with hash code included to FA body.
Do I miss something that might make using of "hashed" bit for FA identification reasonable to consider?

**Comment by vijaysun-omr** [comment]
Thanks @dmitripivkine 

Could you please elaborate on `"hashed" bit might be suddenly be set to 0 by GC` ?

My impression was that the hashed bit gets set when an object gets identity-hashed for the first time and this gets used if/when the object gets moved by the GC. This is because we need the same identity hashcode value to be returned every time for a given object and the value returned the first time an object is identity-hashed is based on the address of the object at that point in time. So when the GC moves the object it needs to grow a slot next to the object to store the original identity hashcode value that has already been returned at least once for that object and use that slot in the future if identity hashcode is called on that object. 

The way it could perhaps work for FA is that we compute the identity hashcode as usual based on its address if it ever gets identity-hashed the first time.  Whether an FA actually gets hashed or not, if it ever gets moved the GC would need to compute the identity hash and add it to the slot that we unconditionally grow when we move the FA. So this becomes a tradeoff : extra overhead when we move a FA in the GC vs extra overhead when we access aaload/aastore in the JITed code (and interpreter). 

I could be missing some subtlety though about how this logic/bits work in the GC.

**Comment by dmitripivkine** [comment]
As I said GC uses "hashed" and "moved" bits as a pair (enum) with states "10","11","01". This is necessary to recognize state "hashed object has been moved with growing **this GC cycle**" to handle Abort cases in Scavenger and Concurrent Scavenger. So state machine for ("hashed","moved") might be "10" --> "01" --> "11" (hashed not moved --> hashed just moved --> hashed and moved). State "hashed just moved" has hash bit set to 0 ("01") obviously.

So if initial state for FA object is going to be "hashed" and "moved" with hash code included right away this might work - in this case GC would treat FA as any other hashed and moved object no exceptions. I don't see any realistic possibility to "tradeoff" dynamic recognition of FA vs hashed objects handling in GC

**Comment by dmitripivkine** [comment]
And if FA object is going to be created "hashed" and "moved" at the beginning how it helps to recognize FA from any other hashed and moved object (array)?

**Comment by DanHeidinga** [comment]
> 3\. Use one of the low bits in `dataAddr` field of the `J9IndexableObject` header to denote a flattened array

Looking at the structures in j9nonbuilder.h, this seems like the only guaranteed free bit.  It'll cost a mask on any data access but as we've seen with masking the class pointers, the overall cost here is pretty close to unmeasurable

**Comment by gacholio** [comment]
2 and 3 are similar solutions - both turn a fetch into a fetch and mask.

I would be tempted to use the size bit to avoid AGI (address generation interlock) penalties when accessing the data. I suspect many checks of the size field can be optimized away by the JIT.

**Comment by gacholio** [comment]
Another plus to using the size field is that all of the VM accesses to it already go through macros.

**Comment by vijaysun-omr** [comment]
The `dataAddr` field is not expected to be used in the JITed code sequence for array access under gencon I believe, as opposed to the array size which is used.

The `dataAddr` field would be used in the JITed code sequence for array access under balanced, but I'd expect many of the cases (though not all) where we commoned the array size to also be cases in which we common the `dataAddr`, e.g. if we versioned a loop, the array size would have to be loop invariant meaning we would have the `dataAddr` also be hoisted out of the loop. Similar argument would probably apply to PRE and local CSE as well. Obviously there are some cases in which the array length can be folded away completely, e.g. through value propagation, but I don't expect that difference to be significant with respect to the choice between `dataAddr` and array size for this discussion on particular.

I think I'd prefer the `dataAddr` option (3) for these reasons and the fact that its still relatively new and so we don't have much JIT code that we'd need to change to account for the extra bit. But I'd be interested in hearing @0xdaryl  view as well.

**Comment by gacholio** [comment]
`dataAddr` is only defined for 64-bit targets.

**Comment by vijaysun-omr** [comment]
> `dataAddr` is only defined for 64-bit targets.

Is this by design and therefore not subject to change ? If the answer to that is "yes" then (3) was'nt much of an option to begin with :)

**Comment by gacholio** [comment]
@bragaigor

**Comment by DanHeidinga** [comment]
> `dataAddr` is only defined for 64-bit targets.

Isn't that effectively all the platforms we care about?

OpenJ9 only builds 32 bit VMs for JDK8 at this point... and maybe IBM still builds 31-bit on zOS for 11?  I'm not sure we care to support 32 bit platforms in the future as we've been actively rejecting requests for them on anything after 8.

**Comment by bragaigor** [comment]
> `dataAddr` is only defined for 64-bit targets.

That's correct (the main reason is to take advantage of the vast 64bit virtual address space)

`dataAddr` was introduced as the primary objective to aid [off-heap technology](https://github.com/eclipse/openj9/issues/11438) in balanced. But now `dataAddr` is in place for all policies and 64bit platforms. If 64 bit constraint is not a problem then `dataAddr` is viable solution as we probably won't be using all 64 bits of this field. Like Dan said, we can just mask away data access in that case. Once we're sure `dataAddr` is in place, one of the plans were to simplify things like `J9JAVAARRAYCONTIGUOUS_EA` where instead of relying on array header size, we could just follow `dataAddr` to access array data, but I think that's still open for discussion.

**Comment by fjeremic** [comment]
For simplicity sake I prefer option (4). How much performance are we actually going to gain with this bit and is it worth the complexity it introduces?

**Comment by 0xdaryl** [comment]
My first instinct when I read the proposed options was to prefer piggybacking on the new `dataAddr` field, mainly due to the fact that it is relatively new with few uses and an investigation has already been conducted to verify the overall minimal impact of introducing it.

When folding the check for the flattened array load (for example), we end up checking both the `J9ClassIsValueType` and the `J9ClassIsFlattened` on the `classFlags`.  Is there value in moving only one of these to the header, or should both move?

> How much performance are we actually going to gain with this bit and is it worth the complexity it introduces?

Our initial, wet-finger experiment should be made more scientific, but for just the flattened array load case performance improved by 2% on x86 by moving the checks into the header rather than digging them out of the `classFlag`.  2% could be important to close the gap between a non-VT and VT build even when VTs are not used.

**Comment by gacholio** [comment]
Does flattened not imply value type?

**Comment by fjeremic** [comment]
> Our initial, wet-finger experiment should be made more scientific, but for just the flattened array load case performance improved by 2% on x86 by moving the checks into the header rather than digging them out of the `classFlag`. 2% could be important to close the gap between a non-VT and VT build even when VTs are not used.

2% on which workload and how heavy was that workload in terms of making use of arrays and specifically value type arrays? Also what is the gap to close between non-VT and VT? Is this 2% a huge portion of the gap?

> Looking at the structures in j9nonbuilder.h, this seems like the only guaranteed free bit. It'll cost a mask on any data access but as we've seen with masking the class pointers, the overall cost here is pretty close to unmeasurable

On Z at least this is close to unmeasurable because we worked with the hardware to introduce an instruction which loads and masks off the low order byte. The `dataAddr` field will not be aligned like `J9Class` pointers in the header are, so we cannot use the "load and mask low order byte" instruction. It will have to be a logical AND operation which as GAC mentions will introduce an AGI for the subsequent load of that address. I'm not sure this will unmeasurable. I suppose we will mostly use `dataAddr` for balanced though so it limits the scope.

**Comment by bragaigor** [comment]
@amicic

**Comment by 0xdaryl** [comment]
The 2% overhead was measured on SPECjbb2005, which doesn't use VT.  It was merely a curiosity experiment while trying to address the throughput gap between builds with VT capabilities built in and non-VT builds.  Ideally, the overhead between the two builds when not running code that doesn't use value types should be zero (as is the case with almost all code we run these days).

Obviously, this overhead can and should be studied on other workloads.
